ethereum_bridge/src/protocol/transactions/ethereum_events/events.rs

```diff
@@ -182,51 +190,66 @@ where
         token::balance_key(&wl_storage.storage.native_token, &BRIDGE_ADDRESS);
     let receiver_native_token_balance_key =
         token::balance_key(&wl_storage.storage.native_token, receiver);
+    let native_werc20_supply_key =
+        minted_balance_key(&wrapped_erc20s::token(native_erc20));
 
-    let eth_bridge_native_token_balance_pre: token::Amount =
-        StorageRead::read(wl_storage, &eth_bridge_native_token_balance_key)?
-            .expect(
-                "Ethereum bridge must always have an explicit balance of the \
-                 native token",
-            );
-    let receiver_native_token_balance_pre: token::Amount =
-        StorageRead::read(wl_storage, &receiver_native_token_balance_key)?
-            .unwrap_or_default();
-
-    let eth_bridge_native_token_balance_post =
-        eth_bridge_native_token_balance_pre
-            .checked_sub(*amount)
-            .expect(
-                "Ethereum bridge should always have enough native tokens to \
-                 redeem any confirmed transfers",
-            );
-    let receiver_native_token_balance_post = receiver_native_token_balance_pre
-        .checked_add(*amount)
-        .expect("Receiver's balance is full");
-
-    StorageWrite::write(
+    update::amount(
         wl_storage,
         &eth_bridge_native_token_balance_key,
-        eth_bridge_native_token_balance_post,
+        |balance| {
+            tracing::debug!(
+                %eth_bridge_native_token_balance_key,
+                ?balance,
+                "Existing value found",
+            );
+            balance.spend(amount);
+            tracing::debug!(
+                %eth_bridge_native_token_balance_key,
+                ?balance,
+                "New value calculated",
+            );
+        },
     )?;
-    StorageWrite::write(
+    update::amount(
         wl_storage,
         &receiver_native_token_balance_key,
-        receiver_native_token_balance_post,
+        |balance| {
+            tracing::debug!(
+                %receiver_native_token_balance_key,
+                ?balance,
+                "Existing value found",
+            );
+            balance.receive(amount);
+            tracing::debug!(
+                %receiver_native_token_balance_key,
+                ?balance,
+                "New value calculated",
+            );
+        },
     )?;
+    update::amount(wl_storage, &native_werc20_supply_key, |balance| {
```

Comment:
The native werc20 supply is exactly the same as the amount of nam escrowed in the ethereum bridge. Unless there is a really compelling reason not to, we shouldn't duplicate state as it will be prone to bugs.

Reply:
it's not quite the same state. the state corresponding to the tokens escrowed under the Ethereum bridge address may be reverted (e.g. if the transfer in the pool times out), while the state corresponding to the total minted supply has been confirmed on Ethereum

Reply:
this follows the pattern with the rest of the minted supply keys we change throughout the code. only state confirmed on Ethereum makes it to storage.